### PR TITLE
Load prequantized bnb 4-bit MoE expert checkpoints under transformers v5

### PR DESCRIPTION
**Pull request by danielhanchen**
## Summary

Older unsloth-bnb-4bit MoE checkpoints store one quantized tensor per expert per projection (`experts.<N>.gate_proj.weight` plus absmax/quant_map/quant_state aux keys), and also quantize small non-Linear weights such as the MoE router. These checkpoints fail to load under transformers v5: the model's MergeModulelist converters match the per-expert keys first, byte-concatenate the packed uint8 as if it were bf16, and drop every aux key as unexpected, while the router lands as raw packed bytes on a plain `nn.Parameter` slot and `F.linear` fails with a reduction dim mismatch.

## What this does

All three patches are active only when a `Bnb4BitHfQuantizer` is `pre_quantized`:

1. `patch_bnb4bit_model_conversion_mapping` prepends quantized twins of the model's per-expert merge converters. Each twin collects the aux keys, rebuilds one `QuantState` per expert, concatenates the packed bytes in target element order (NF4 packing is elementwise row-major, so this is exact), and reassembles a single stacked `Params4bit` with a flat fp32 absmax. Bare weight patterns are anchored so the twins never shadow the model converters for bf16 checkpoints, and layers without aux keys (dynamic-quant skip layers) fall back to the model's own merge.
2. `patch_bnb4bit_quantizer_weight_conversions` registers deserializers for fused expert param names (`gate_up_proj`/`down_proj`), which the stock quantizer only registers for params literally named `weight`.
3. `patch_bnb4bit_dequantize_plain_params` dequantizes any `Params4bit` that landed on a module that is neither a bnb `Linear4bit` nor a v5 experts module (e.g. the router) back to a plain float Parameter after loading.

## Testing

- Dequantized expert weights verified bitwise identical to a reference load.
- End-to-end 4-bit LoRA training validated on Qwen3-30B-A3B (finite losses, 4-bit resident memory).
- 5 new CPU unit tests cover twin construction, pattern anchoring, the unquantized-layer fallback, and absmax denesting.
- Validated across transformers 4.57.6, 5.5, and 5.13 (the last routes stacked experts through PEFT target_parameters, which needs the logical 3D shape now set on the reassembled param).

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
Addressed the review points: set `_original_shape` on the deserialized `Params4bit` so PEFT can read it, filtered the per-expert stack deserialize to only the sources actually present in the checkpoint (no more KeyError/IndexError on a partial mapping), guarded `_process_model_after_weight_loading` with `getattr` for transformers versions that lack it, copied `input_dict` before mutating, and preserved each parameter's `requires_grad` on dequant. The trailing `pass` and `return raise_error(...)` style match the rest of the temporary_patches modules, so I left those. Tests pass.

File: unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py [hunk 1]
```diff
@@ -586,12 +589,353 @@ def _patched_unmerge(self):
...
+    if original_get_weight_conversions is None:
+        return
+    if getattr(original_get_weight_conversions, "_unsloth_moe_patched", False):
+        return
+
+    def patched_get_weight_conversions(self):
+        conversions = original_get_weight_conversions(self)
+        if self.pre_quantized:
+            try:
+                conversions = list(conversions) + _bnb4bit_expert_weight_conversions(self)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py:668]: Good catch, this was real for models like qwen3_vl_moe whose fused converter precedes the quantizer conversions. Added prepended fused twins that deserialize when aux keys are present and replicate the original ops otherwise.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py [hunk 2]
```diff
@@ -586,12 +586,448 @@ def _patched_unmerge(self):
...
+                if (self.base_source + suf) in input_dict
+            }
+            module, _ = get_module_from_name(model, full_layer_name)
+            new_value = Params4bit.from_prequantized(
+                data=weight,
+                quantized_stats=stats,
+                requires_grad=False,
+                device=weight.device,
+                module=module,
+            )
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py:901]: Converters run before the quantization op in core loading, so a prequantized fused checkpoint produced by save_pretrained is always in the model layout with that layout baked into the quant stats; on save the reverted check_dims Transpose compares the packed param against itself and never fires. A transpose also cannot be applied to packed 4-bit data, so honoring the op here is not possible. Added a guard that raises a clear error if a checkpoint quantized in the transposed layout ever shows up.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py [hunk 3]
```diff
@@ -586,12 +586,468 @@ def _patched_unmerge(self):
...
+    )
+pass
+
+
+def _dequantize_plain_param_slots(model: nn.Module):
+    if not HAS_BNB:
+        return
+    for module_name, module in model.named_modules():
+        if isinstance(module, (bnb.nn.Linear4bit,)) or _is_expert_module(module):
+            continue
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py]: Good catch. The pass now skips every bitsandbytes module (checked via the class MRO), so Embedding4bit and friends keep their quantized weights; only genuinely plain parameter slots convert. Added a test.

File: unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py [hunk 4]
```diff
@@ -586,12 +586,468 @@ def _patched_unmerge(self):
...
+            if any(s[1] != in_dim for s in src_shapes):
+                raise ValueError(f"Unsloth: mismatched expert in_dims {src_shapes} for {full_layer_name}")
+
+            packed_rows, absmax_rows = [], []
+            for e in range(num_experts):
+                packed_rows.append(torch.cat(
+                    [input_dict[anch][e].reshape(-1) for _, anch in present_sources]
+                ))
+                absmax_rows.append(torch.cat([per_src_absmax[i][e] for i in range(len(present_sources))]))
+            data = torch.stack(packed_rows).unsqueeze(-1)  # (E, bytes_per_expert, 1)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py]: Confirmed with a repro: a 60-element segment against blocksize 64 mis-scales everything after the partial block. The rebuild now checks alignment and repacks from dequantized values when any segment is not a blocksize multiple; raw concat is kept for the aligned case where it is byte-exact. Added parametrized tests for both paths.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py [hunk 5]
```diff
@@ -586,12 +586,503 @@ def _patched_unmerge(self):
...
+            if UNSLOTH_ENABLE_LOGGING:
+                logger.info(f"Unsloth: plain-param dequantize pass failed: {e}")
+        return result
+
+    patched_after_load._unsloth_plain_dequant_patched = True
+    patch_function(
+        Bnb4BitHfQuantizer, "_process_model_after_weight_loading", patched_after_load,
+        match_level="relaxed",
+    )
+pass
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py:1043]: The trailing pass after patch_function is the separator style used throughout the temporary_patches modules (11 in this file alone), so I am keeping it for consistency.

File: unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py [hunk 6]
```diff
@@ -586,12 +589,353 @@ def _patched_unmerge(self):
...
+            new_param.module = module
+            module._is_hf_initialized = True
+            return {target_patterns[0]: new_param}
+
+    twins = []
+    for conv in model_conversions:
+        if not isinstance(conv, WeightConverter):
+            continue
+        target = conv.target_patterns[0]
+        if not target.endswith(("gate_up_proj", "down_proj")):
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py:937]: This path is now restricted to single-target converters: the loop skips any converter whose target_patterns is not length 1, so a many-to-many mapping keeps its original conversion and no expert target is left unloaded.

File: unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py [hunk 7]
```diff
@@ -586,12 +589,353 @@ def _patched_unmerge(self):
...
+                    input_dict[key] = value[0]
+            weight = input_dict.pop(self.param_name)
+            module, _ = get_module_from_name(model, full_layer_name)
+            new_value = Params4bit.from_prequantized(
+                data=weight,
+                quantized_stats=input_dict,
+                requires_grad=False,
+                device=weight.device,
+                module=module,
+            )
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/moe_utils_bnb4bit.py:624]: The fused deserializer now records the logical 3D shape from the quant state (new_value._original_shape = new_value.quant_state.shape), so PEFT LoRA can recover it for prequantized fused checkpoints.

**Comment by danielhanchen** [comment]
@codex review
